### ISSUE TITLE
stability: single slow replica can muck up Raft processing

### ISSUE THREAD

**Issue by petermattis** [cockroachdb/cockroach#8638]
Applying a 64MB snapshot for a replica can take over a second. The current structure of `Store.processRaft` allows replicas to be processed concurrently, but only a singe step at a time. So when we apply a snapshot to a replica, all the other replicas on the store experience a hiccup where we don't process their requests and don't tick their raft groups. This can lead to undesirable raft elections and user visible performance hiccups.

Also see #8548 which talks about `applySnapshot` efficiency. But even if we make `applySnapshot` faster we'll still have problems with the raft processing queue getting bottlenecked by a snapshot being applied.

I'm not sure if streaming snapshots addresses this problem. Even if it does, it seems undesirable for one bad replica to foul up all of the replicas on a store.

@bdarnell, @spencerkimball Thoughts?

**Comment by spencerkimball** [comment]
@cockroachdb/stability

**Comment by bdarnell** [comment]
The processRaft loop was designed with the assumption that everything we do would be shorter than the raft tick interval (even applying snapshots - 64MB shouldn't take too long, although our current tick interval is probably too short for this case). In the short term, we should increase the tick interval so we don't skip ticks due to snapshot processing. 

In the longer term, we should probably rework the processRaft goroutine so that it doesn't process all the replicas in lockstep. The trickiest parts of this are in dealing with the uninitialized replicas (which are currently handled as a special case before initialized replicas are handed off to the pool of workers) and synchronization between RemoveReplica and raft processing (which is the reason we have the coarse-grained processRaftMu).

**Comment by petermattis** [comment]
From discussion on [#8630](https://github.com/cockroachdb/cockroach/issues/8630#issuecomment-240898505):

Here's a partially baked idea: we replace `Store.processRaftMu` with `Replica.raftMu` and `Store.uninitRaftMu`. Most operations just require holding `Replica.raftMu`. The exception is splits which also need to grab the mutex for the right-side of the split. The primary difficulty here is that an uninitialized replica doesn't know the key range it applies to. Therefore `Store.uninitRaftMu` would be locked whenever performing raft related activity on an uninitialized replica. Splits would grab `Replica.raftMu` for both the left and right hand side range as well as `Store.uninitRaftMu`. 

Similar to the current setup, uninitialized replicas (including preemptive snapshots) would be processed serially. Initialized replicas could be processed (and ticked) concurrently and independently of the single current `Store.processRaft` loop. I think I would try to restructure that processing so that a pool of raft workers independently picks up and performs work for either a single initialized replica or the set of all uninitialized replicas. 

This would be a big, hairy, scary change to the locking. I'll let it percolate for a bit.

**Comment by petermattis** [comment]
Any thoughts on my proposal above? I've explored it a bit and it seems workable. In addition to changing the locking we would have to change how incoming `RaftMessages` are processed for uninitialized replicas so that it doesn't block the goroutine reading the messages from the stream.

**Comment by bdarnell** [comment]
Sounds reasonable to me, except the lock ordering for splits is unclear. It sounds like you want to hold uninitReplicaMu before any Replica.raftMu when ticking uninitialized replicas, but splits would presumably start by holding the left Replica.raftMu. I suppose it's not technically a lock ordering violation because the lock you hold before uninitReplicaMu is for an initialized replica, but it still seems tricky. Or did you have another ordering in mind?

**Comment by tschottdorf** [comment]
One fun way to stress this kind of issue locally is

```bash
echo 'create table if not exists blobs.blobs(blob bytes);' | ./cockroach sql
while [ $? -eq 0 ]; do sleep 1; echo 'insert into blobs.blobs values(repeat(now()::string, power(2, 20::float)::int));' | ./cockroach sql; done
```

This essentially inserts 32mb blobs in random locations. It's not representative of any kind of workload we want to support, but it does give a fairly good approximation of (non-streaming) snapshots and in particular the cost which is eaten in applying that large request all at once.

It's also a good way to create a lot of large ranges locally; at least every second insert will cause a split.

I think at the end of this refactoring, memory permitting, we should be handling this load much better. The Raft timing graphs show clearly how the goroutine is struggling to keep up, and without the `sleep 1` gets quickly overwhelmed even though this is only a single threaded insert.

**Comment by petermattis** [comment]
Ok, now that #8941 has the locking changed it is time to start thinking about how to break up `Store.processRaft` and `Store.HandleRaftRequest`. Currently, handling of a preemptive snapshot will block other incoming requests from the remote node. I think we should queue up the preemptive snapshot on the uninitialized replica and let the raft processing goroutine handle it. And we should split `Store.processRaft` into two goroutines: one for initialized replicas and one for uninitialized ones. Together, these two changes would allow initialized replicas to continue processing when a preemptive snapshot is being applied.

The above approach would still suffer on Raft-initiated snapshots. Such snapshots could block processing of other initialized replicas. I think a more significant restructuring of `Store.processRaft` would be needed to address this, but perhaps someone else has idea here.

**Comment by jordanlewis** [comment]
Is there any chance that this thinking could be placed on hold until we merge streaming snapshots into master? I'm nervous about the merge, and I think if we do want to go in the direction of streaming snapshots it would be better not to duplicate this thinking work pre- and post- streaming snapshots.

**Comment by petermattis** [comment]
Placing this thinking on hold is a bit aggressive, but we should certainly discuss the sequencing of when these changes merge vs streaming snapshots. Streaming snapshots are important and tricky, but I'd argue that the locking changes in #8941 are trickier. FWIW, I'd rather rewrite streaming snapshots on top of #8941 than rewrite #8941 on top of streaming snapshots. Let's talk about it next Tuesday.

**Comment by tamird** [comment]
Unfortunately, that means that merging changes back from master to develop
is blocked until a decision can be reached.

**Comment by jordanlewis** [comment]
FWIW my question above was more about sequencing the next step changes you proposed above. I agree that streaming snapshots needs to come on top of #8941.

**Comment by petermattis** [comment]
Ok. Let's talk about that next week. The changes I'm describing above are much more straightforward than #8941.

**Comment by bdarnell** [comment]
With processRaftMu gone, I think we can get rid of the WaitGroup in processRaft: just pass ready replicas off to the workers and return to the select loop without waiting for them to finish. Then activity one one replica doesn't have to block activity on others (unless we run out of goroutines in the worker pool).

**Comment by petermattis** [comment]
Good point about not needing the WaitGroup for initialized replicas.

**Comment by petermattis** [comment]
After #8941 goes in, is there any reason to distinguish initialized vs uninitialized replicas in `Store.processRaft`?

Streaming snapshots will move processing of snapshots off the `HandleRaftRequest` goroutine (each stream will start a new goroutine on the recipient), but we'll still want to move `RaftMessageRequest` processing off of the `HandleRaftRequest` goroutines (there is one per remote store) so that receiving a heartbeat request for a replica that has a snapshot being applied does not block reception of requests to other replicas. 

A test that I'd like to see written before this issue is closed: 

1.  Set up 2 node cluster containing 2 ranges.
2. Grab the `*Replica` for one of the ranges on one of the nodes and lock `Replica.raftMu`. 
3. Verify that we can send raft requests and heartbeats for the other range.

**Comment by bdarnell** [comment]
I don't think so; `uninitRaftMu` should completely replace the special treatment of uninitialized replicas in `processRaft`.

**Comment by petermattis** [comment]
The next task: `Store.HandleRaftRequest` grabs `Replica.raftMu`, but doing so blocks processing of messages from other Replicas from the same remote store (there is a single goroutine receiving messages from each remote store). In discussion with @tschottdorf and @spencerkimball, my plan is to add `Store.mu.replicaQueues[roachpb.RangeID][]*RaftMessageRequest`. When a message arrives, we add the message to the associated replica queue and kick a pool of goroutines to pick up the replica for processing. 

We use a per-replica queue so that multiple messages for the same replica can't consume more than 1 worker goroutine. We use a queue that is external from `Replica` so that I don't have to revisit the locking of `Store.getOrCreateReplica` which returns a replica with `Replica.raftMu` already locked.

The size of the per-replica queues is TBD. So is the behavior when the queue is full. We can simply drop messages, but eventually we'll want to provide back pressure to the remote side. The bidirectional stream gives a mechanism for doing this.
